active/0004-integrating-extensible-resources-and-deployment-stacks.md

```diff
@@ -0,0 +1,547 @@
+
+### OAuth 2.0 On-behalf-of (OAuth OBO) flow
+
+The [Microsoft identity platform and OAuth 2.0 On-Behalf-Of flow](https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-on-behalf-of-flow), which describes the scenario of a web API using an identity other than its own to call another web API. Referred to as delegation in OAuth, the intent is to pass a user's identity and permissions through the request chain.
+
+The flow is a non-standard OAuth 2.0 extension developed by Microsoft.
+
+## Motivation
+
+Deployment Stacks offer users the convenience of managing a set of resources as a unified entity, enabling quicker updates and deletions. Currently, this functionality is limited to Azure resources due to the absence of a unique identifier (resource ID) for extensible resources and a standardized authentication mechanism across different extensibility providers.
+
+To overcome these limitations, the REP suggests modifications to the deployments API. A proposed change is the introduction of a `referenceID` field in `outputResources`, a freefrom string designed to uniquely identify extensible resources. Additionally, the proposal includes updates to the `deploymentProviders`, enabling it to store provider configurations. These configurations may include references to key vaults or resource properties, which can be utilized to access authentication details safely. These proposed enhancements allows Deployment Stacks to interact with extensibility providers to delete extensible resources.
+
+Implementing these changes ensure that extensible resources can be managed with the same level of efficiency and security as Azure resources within Deployment Stacks.
+
+## Detailed design
+
+### Reference ID
+
+> A reference ID is a URL-encoded, free-form string designed to uniquely identify an extensible resource. This ID is generated by extensibility providers and is returned as part of the responses from the `POST /resources/reference` or `POST /resources/preview` extensible resource APIs. The concept and design of reference IDs have been finalized.
+>
+> Any reference ID MUST adhere to the format `{flags}-{uniqueId}`.
+
+> - `{flags}`: This section consists of two binary digits indicating the type of reference ID. The currently accepted values are `00` and `01`:
+>   - `00`: signifies that the reference ID is autonomous and doesn't need extra data for GET and DELETE actions. This implies the utilization of the extensible resource APIs `GET /resources/{referenceId}` and `DELETE /resources/{referenceId}` for operations.
+>   - `01`: indicates that additional information, specifically user-provided provider configuration (like a connection string), is essential alongside the reference ID to execute GET and DELETE operations. In this case, APIs `POST /resources/{referenceId}/getWithConfig` and `POST /resources/{referenceId}/deleteWithConfig` must be used, and the provider configuration must be included in the request body.
```

Comment:
Apologies if I missed a previous discussion on this ...

Have you considered turning this into a more RPC-style API? The trailing operation name (eg: `/deleteWithConfig` at the end of the URL path) is hard to parse in an unambiguous way. 

ARM has rules to disambiguate this, but I'm not sure this will be true of all reference ID formats. In any case where the reference ID has a variable number of segments, this is pretty difficult to handle, and your web framework can't help.

For an example in ASP.NET core, you can write a route to handle `/resources/{**referenceId}`, but you can't write a route to handle `/resources/{**referenceId}/deleteWithConfig` (it's blocked by the routing engine). Most routing frameworks have the same limitation.

Reply:
Is this true if `referenceId` even if it is URL-encoded?

Reply:
If the API call is going to include a URL-encoded string **inside** a path segment like `/resources/<url-encoded-text-here>/deleteWithConfig` then you're going to see inconsistent behaviors across webservers when that segment contains `%2F` (encoded text for `/`) or path navigation operators like `..` which don't require encoding.
 
For example I believe ASP.NET Core (using it as an example because I know it well) behaves differently depending on whether IIS is in front of it or not.
 
Servers take differing approaches to cases like these because URL paths often map to authorization rules and other security features. The risk that they are mitigating is when the server code (or middleware) interprets the URL differently than application code.
 
If uniqueId is a free-form string then passing it in the request body is the safest option. This is already an RPC-style API, so I'm not sure why it's valuable to pass the uniqueId in the URL.
 
Let me know how I can help, this design feels like it could be simplified.

Reply:
This is good to know! The other comments and questions also make sense. As discussed offline, the v2 APIs require a revision. Will update this PR later to incorporate the changes once the [new extensibility v2 API design](https://github.com/Azure/bicep-extensibility/pull/176) is approved.